netdata/netdata
# Gap Filling in a parent node for data coming from replicated child nodes

##### Summary

The goal of this PR is to be able to fill gaps of a replicated child Netdata agent on a parent Netdata agent following a network failure or a parent restart in a simple and fast solution with low complexity.

##### Terminology: 
* `stream replication` is the tech term for the `gap filling` feature. 
* `hard gaps` are the gaps that neither Netdata agent has data for it. These gaps happening when the source Netdata agents are down and not collecting data.
* `soft gaps` are the gaps that the child Netdata agent has data to replicate to the parent side in order to complete the gap. 

##### Test Plan

Set up a `parent <-> child` architecture and run the end-user scenarios below,
 1. Start the parent, start the child. Disconnect the network. Reconnect the network. - Expected result: `soft gap` filling.
 2. Start the parent, start the child. Shutdown the parent. Start the parent. - Expected result: `soft gap` filling.
 3. Start the child. Start the parent. Repeat 1.-2. with this starting sequence. - Expected result: `hard gap` not filled and `soft gap` filling.
 
 On both agents `netdata.conf `
 ```
 [global]
    ...
    memory mode = dbengine
    debug flags = 0x0000001000000000
    ...
 ```
 Child `stream.conf`
 ```
 [stream]
    # Enable this on child nodes, to have them send metrics.
    enabled = yes
    # Enable replication
    enable replication = yes
    ...
 ```
 Parent `stream.conf`
 ```
 [stream]
    ...
    # Enable replication
    enable replication = no
    ...
 [api-key-api-key-apikey]
    enabled = yes
    enable replication = yes
 ```
#### Limitations & Known issues
* `hard gaps` cannot be filled since there are not metrics in any Netdata agent to replicate them.
* At the moment, the parent can handle a maximum gap length of replicated data for up to 17 minutes.
* At the moment, gap filling is supported on memory mode `dbengine` for parent agents and `dbengine/ram` mode for child agents.

<details> <summary>For users: How does this change affect me?</summary>
  
The gap filling feature is the completion of the metrics gaps in the parent's Netdata dashboard when the missing metrics are available in the child Netdata agents. The gaps on the Netdata dashboard metrics can happen because of short network disconnections, Netdata agent restarts or a machine failure in the parent side. Enabling the gap filling feature allows to the parent Netdata agent to report the gaps to the child Netdata agent and receive the missing metrics to fill the gaps.
</details>
